File: libs/java/auth_core/src/main/java/com/yahoo/athenz/auth/impl/SimplePrincipal.java
/*
 * Copyright The Athenz Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.yahoo.athenz.auth.impl;

import java.security.cert.X509Certificate;
import java.util.List;
import java.util.Objects;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.yahoo.athenz.auth.Authority;
import com.yahoo.athenz.auth.Principal;

public class SimplePrincipal implements Principal {

    private static final Logger LOG = LoggerFactory.getLogger(SimplePrincipal.class);
    
    String domain;
    String name = null;
    String fullName  = null;
    String creds;
    String unsignedCreds = null;
    String ip = null;
    long issueTime = 0;
    List<String> roles = null; 
    Authority authority;
    String authorizedService = null;
    String originalRequestor = null;
    String keyService = null;
    String keyId = null;
    X509Certificate x509Certificate = null;
    String applicationId = null;
    private boolean mtlsRestricted = false;
    String rolePrincipalName = null;

    // defaulting to ACTIVE state
    private Principal.State state = State.ACTIVE;

    public static Principal create(String domain, String name, String creds) {
        return create(domain, name, creds, 0, null);
    }

    /**
     * Create a Principal based on a given Access/Role Token or role certificate
     * @param domain Domain name that the RoleToken was issued for
     * @param creds Credentials of the principal (RoleToken)
     * @param roles List of roles defined in the token
     * @param authority authority responsible for the credentials (e.g. RoleAuthority)
     * @return a Principal for the given set of roles in a domain
     */
    public static Principal create(String domain, String creds, List<String> roles, Authority authority) {
        return create(domain, creds, roles, null, authority);
    }

    /**
     * Create a Principal based on a given Access/Role Token or role certificate
     * @param domain Domain name that the RoleToken was issued for
     * @param creds Credentials of the principal (RoleToken)
     * @param roles List of roles defined in the token
     * @param rolePrincipalName principal who requested the given access token or role certificate
     * @param authority authority responsible for the credentials (e.g. RoleAuthority)
     * @return a Principal for the given set of roles in a domain
     */
    public static Principal create(String domain, String creds, List<String> roles, String rolePrincipalName, Authority authority) {
        if (roles == null || roles.isEmpty()) {
            LOG.error("createRolePrincipal: zero roles");
            return null;
        }
        return new SimplePrincipal(domain, creds, roles, rolePrincipalName, authority);
    }

    /** 
     * Create a Principal for the given identity
     * @param domain Domain name for the identity
     * @param name Name of the identity
     * @param creds Credentials of the principal (PrincipalToken which could be either UserToken or ServiceToken)
     * @param authority authority responsible for the credentials (e.g. PrincipalAuthority)
     * @return a Principal for the identity
     */
    public static Principal create(String domain, String name, String creds, Authority authority) {
        return create(domain, name, creds, 0, authority);
    }

    /** 
     * Create a Principal for the given user identity
     * @param domain Domain name for the identity (For users this will always be user)
     * @param name Name of the identity
     * @param creds Credentials of the principal (e.g. Cookie.User)
     * @param issueTime when the User Cookie/Credentials was issued
     * @param authority authority responsible for the credentials (e.g. UserAuthority)
     * @return a Principal for the identity
     */
    public static Principal create(String domain, String name, String creds, long issueTime, Authority authority) {
        String matchDomain = (authority == null) ? null : authority.getDomain();
        if (matchDomain != null && !domain.equals(matchDomain)) {
            LOG.error("createPrincipal: domain mismatch for user {} in authority {}", name, authority);
            return null;
        }
        return new SimplePrincipal(domain, name, creds, issueTime, authority);
    }

    /** 
     * Create a Principal for the given host identity
     * @param appId Application identifier
     * @param creds Credentials of the principal
     * @param authority authority responsible for the credentials (e.g. HostAuthority)
     * @return a Principal for the host identity
     */
    public static Principal create(String appId, String creds, Authority authority) {
        if (appId == null) {
            LOG.error("createAppIdPrincipal: null appId");
            return null;
        }
        return new SimplePrincipal(null, appId, creds, 0, authority);
    }
    
    private SimplePrincipal(String domain, String name, String creds, long issueTime, Authority authority) {
        this.domain = domain;
        this.name = name;
        this.creds = creds;
        this.authority = authority;
        this.issueTime = issueTime;
    }

    private SimplePrincipal(String domain, String creds, List<String> roles, String rolePrincipalName, Authority authority) {
        this.domain = domain;
        this.creds = creds;
        this.roles = roles;
        this.authority = authority;
        this.rolePrincipalName = rolePrincipalName;
    }

    public void setUnsignedCreds(String unsignedCreds) {
        this.unsignedCreds = unsignedCreds;
    }
    
    public void setAuthorizedService(String authorizedService) {
        this.authorizedService = authorizedService;
    }
    
    public void setIP(String ip) {
        this.ip = ip;
    }
    
    public void setOriginalRequestor(String originalRequestor) {
        this.originalRequestor = originalRequestor;
    }
    
    public void setKeyService(String keyService) {
        this.keyService = keyService;
    }
    
    public void setKeyId(String keyId) {
        this.keyId = keyId;
    }
    
    public void setX509Certificate(X509Certificate x509Certificate) {
        this.x509Certificate = x509Certificate;
    }
    
    public void setApplicationId(String applicationId) {
        this.applicationId = applicationId;
    }
    
    public void setRoles(List<String> roles) {
        this.roles = roles;
    }

    public void setMtlsRestricted(boolean isMtlsRestricted) {
        this.mtlsRestricted = isMtlsRestricted;
    }

    public void setState(State state) {
        this.state = state;
    }

    public void setRolePrincipalName(String rolePrincipalName) {
        this.rolePrincipalName = rolePrincipalName;
    }

    @Override
    public String getIP() {
        return ip;
    }
    
    @Override
    public String getUnsignedCredentials() {
        return unsignedCreds;
    }
    
    @Override
    public Authority getAuthority() {
        return authority;
    }

    @Override
    public String getDomain() {
        return domain;
    }

    @Override
    public String getName() {
        return name;
    }

    @Override
    public String getOriginalRequestor() {
        return originalRequestor;
    }
    
    @Override
    public String getFullName() {
        
        if (fullName == null) {
            if (domain != null && name != null) {
                fullName = domain + "." + name;
            } else {
                fullName = (name != null) ? name : domain;
            }
        }
        
        return fullName;
    }
    
    @Override
    public String getCredentials() {
        return creds;
    }
    
    @Override
    public List<String> getRoles() {
        return roles;
    }

    @Override
    public long getIssueTime() {
        return issueTime;
    }
    
    public String toString() {
        if (roles == null) {
            return domain + "." + name;
        } else {
            return "ZToken_" + domain + "~" + roles.toString().replace("[", "").replace("]", "");
        }
    }

    @Override
    public String getAuthorizedService() {
        return authorizedService;
    }
    
    @Override
    public String getKeyService() {
        return keyService;
    }
    
    @Override
    public String getKeyId() {
        return keyId;
    }

    @Override
    public X509Certificate getX509Certificate() {
        return x509Certificate;
    }
    
    @Override
    public String getApplicationId() {
        return this.applicationId;
    }

    @Override
    public boolean getMtlsRestricted() {
        return this.mtlsRestricted;
    }

    @Override
    public Principal.State getState() {
        return this.state;
    }

    @Override
    public String getRolePrincipalName() {
        return this.rolePrincipalName;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        SimplePrincipal that = (SimplePrincipal) o;
        return getFullName().equals(that.getFullName());
    }

    @Override
    public int hashCode() {
        return Objects.hash(getFullName());
    }
}

File: libs/nodejs/auth_core/src/impl/SimplePrincipal.js
/**
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
'use strict';

const logger = require('../../logger');
var Validate = require('../util/Validate');
var config = require('../../config/config')();

class SimplePrincipal {
    /*eslint max-params: ["error", 6]*/
    constructor(domain, name, creds, roles, issueTime, authority) {
        this._domain = domain;
        this._name = name;
        this._creds = creds;
        this._roles = roles;
        this._issueTime = Number(issueTime);
        this._authority = authority;
        this._fullName = null;
        this._unsignedCreds = null;
        this._ip = null;
        this._authorizedService = null;
        this._originalRequestor = null;
        this._keyService = null;
        this._keyId = null;
        this._x509Certificate = null;
    }

    static setConfig(c) {
        config = Object.assign({}, config, c.auth_core);
    }

    static _simplePrincipal(domain, name, creds, issueTime, authority) {
        return new SimplePrincipal(
            domain,
            name,
            creds,
            null,
            issueTime,
            authority
        );
    }

    static _simplePrincipalByRoles(domain, creds, roles, authority) {
        return new SimplePrincipal(domain, null, creds, roles, 0, authority);
    }

    static create(domain, name, creds) {
        return this.createByUserIdentity(domain, name, creds, 0, null);
    }

    /**
     * Create a Principal based on a given RoleToken
     * @param domain Domain name that the RoleToken was issued for
     * @param creds Credentials of the principal (RoleToken)
     * @param roles List of roles defined in the token
     * @param authority authority responsible for the credentials (RoleAuthority)
     * @return a Principal for the given set of roles in a domain
     */
    static createByRoles(domain, creds, roles, authority) {
        if (!Validate.domainName(domain)) {
            logger.warn('createByRoles: failed to validate domain ' + domain);
        }

        if (!roles || roles.length === 0) {
            logger.warn('zero roles: ' + creds);
        }

        return this._simplePrincipalByRoles(domain, creds, roles, authority);
    }

    /**
     * Create a Principal for the given identity
     * @param domain Domain name for the identity
     * @param name Name of the identity
     * @param creds Credentials of the principal (PrincipalToken which could be either UserToken or ServiceToken)
     * @param authority authority responsible for the credentials (e.g. PrincipalAuthority)
     * @return a Principal for the identity
     */
    static createByIdentity(domain, name, creds, authority) {
        return this.createByUserIdentity(domain, name, creds, '0', authority);
    }

    /**
     * Create a Principal for the given user identity
     * @param domain Domain name for the identity (For users this will always be user)
     * @param name Name of the identity
     * @param creds Credentials of the principal (e.g. Cookie.User)
     * @param issueTime when the User Cookie/Credentials was issued
     * @param authority authority responsible for the credentials (e.g. UserAuthority)
     * @return a Principal for the identity
     */
    static createByUserIdentity(domain, name, creds, issueTime, authority) {
        if (!Validate.domainName(domain)) {
            logger.warn(
                'createByUserIdentity: failed to validate domain ' + domain
            );
        }

        if (!Validate.principalName(name)) {
            logger.warn(
                'createByUserIdentity: failed to validate name ' + name
            );
        }

        if (domain) {
            var matchDomain = !authority ? null : authority.getDomain();
            if (matchDomain && domain !== matchDomain) {
                logger.warn(
                    'domain mismatch for user ' +
                        name +
                        ' in authority + ' +
                        authority
                );
                return null;
            }
        } else if (authority) {
            if (authority.getDomain()) {
                logger.warn(
                    'domain mismatch for user ' +
                        name +
                        ' in authority + ' +
                        authority
                );
                return null;
            }
        }

        return this._simplePrincipal(domain, name, creds, issueTime, authority);
    }

    /**
     * Create a Principal for the given host identity
     * @param appId Application identifer
     * @param creds Credentials of the principal
     * @param authority authority responsible for the credentials (e.g. HostAuthority)
     * @return a Principal for the host identity
     */
    static createByHostIdentity(appId, creds, authority) {
        return this._simplePrincipal(null, appId, creds, 0, authority);
    }

    setUnsignedCreds(unsignedCreds) {
        this._unsignedCreds = unsignedCreds;
    }

    setAuthorizedService(authorizedService) {
        this._authorizedService = authorizedService;
    }

    setIP(ip) {
        this._ip = ip;
    }

    setOriginalRequestor(originalRequestor) {
        this._originalRequestor = originalRequestor;
    }

    setKeyService(keyService) {
        this._keyService = keyService;
    }

    setKeyId(keyId) {
        this._keyId = keyId;
    }

    setX509Certificate(x509Certificate) {
        this._x509Certificate = x509Certificate;
    }

    getIP() {
        return this._ip;
    }

    getUnsignedCreds() {
        return this._unsignedCreds;
    }

    getAuthority() {
        return this._authority;
    }

    getDomain() {
        return this._domain;
    }

    getName() {
        return this._name;
    }

    getOriginalRequestor() {
        return this._originalRequestor;
    }

    getFullName() {
        if (!this._fullName) {
            if (this._domain && this._name) {
                this._fullName = this._domain + '.' + this._name;
            } else if (this._domain) {
                this._fullName = this._domain;
            } else if (this._name) {
                this._fullName = this._name;
            }
        }
        return this._fullName;
    }

    getCredentials() {
        return this._creds;
    }

    getRoles() {
        return this._roles;
    }

    getIssueTime() {
        return this._issueTime;
    }

    toString() {
        if (!this._roles) {
            return this._domain + '.' + this._name;
        } else {
            return (
                'ZToken_' +
                this._domain +
                '~' +
                this._roles.toString().replace('[', '').replace(']', '')
            );
        }
    }

    getAuthorizedService() {
        return this._authorizedService;
    }

    getKeyService() {
        return this._keyService;
    }

    getKeyId() {
        return this._keyId;
    }

    getX509Certificate() {
        return this._x509Certificate;
    }
}

module.exports = SimplePrincipal;
